### PR TITLE
Change the installer cmd to cobra

### PR DESCRIPTION
**Pull request by marccampbell**

### CODE REVIEW

File: cmd/installer/main.go [hunk 1]
```diff
@@ -0,0 +1,62 @@
...
+	cmd.AddCommand(RestoreCmd(ctx, name))
+	cmd.AddCommand(AdminConsoleCmd(ctx, name))
+	cmd.AddCommand(SupportBundleCmd(ctx, name))
+
+	return cmd
+}
+
+func InitAndExecute(ctx context.Context, name string) {
+	if err := RootCmd(ctx, name).Execute(); err != nil {
+		fmt.Println(err)
```
Comment (emosbaugh) [inline cmd/installer/main.go:59]: this will print to stdout and it should go to stderr

File: cmd/installer/cli/cidr.go [hunk 2]
```diff
@@ -0,0 +1,44 @@
...
+	podCIDRFlag, err := cmd.Flags().GetString("pod-cidr")
+	if err != nil {
+		return "", "", fmt.Errorf("unable to get pod-cidr flag: %w", err)
+	}
+	serviceCIDRFlag, err := cmd.Flags().GetString("service-cidr")
+	if err != nil {
+		return "", "", fmt.Errorf("unable to get service-cidr flag: %w", err)
+	}
+
+	if podCIDRFlag != "" && serviceCIDRFlag != "" {
```
Comment (sgalsaleh) [inline cmd/installer/cli/cidr.go:25]: this needs to be an `||`
```suggestion
	if podCIDRFlag != "" || serviceCIDRFlag != "" {
```

File: cmd/installer/cli/install.go [hunk 3]
```diff
@@ -0,0 +1,1007 @@
...
+
+			defer tryRemoveTmpDirContents(provider)
+
+			var err error
+			err = configutils.WriteRuntimeConfig(runtimeConfig)
+			if err != nil {
+				return fmt.Errorf("unable to write runtime config: %w", err)
+			}
+
+			proxy, err := getProxySpecFromFlags(cmd)
```
Comment (sgalsaleh) [inline cmd/installer/cli/install.go:159]: isn't this basically already done in `PreRunE`? :)

File: cmd/installer/cli/install.go [hunk 4]
```diff
@@ -0,0 +1,1007 @@
...
+			}
+
+			logrus.Debugf("checking license matches")
+			license, err := getLicenseFromFilepath(licenseFlag)
+			if err != nil {
+				metricErr := fmt.Errorf("unable to get license: %w", err)
+				metrics.ReportApplyFinished(cmd.Context(), licenseFlag, metricErr)
+				return err // do not return the metricErr, as we want the user to see the error message without a prefix
+			}
+			isAirgap := false
```
Comment (sgalsaleh) [inline cmd/installer/cli/install.go:221]: nit:
```suggestion
			isAirgap := airgapBundle != ""
```

File: cmd/installer/cli/install.go [hunk 5]
```diff
@@ -0,0 +1,1007 @@
...
+				isAirgap = true
+			}
+			if isAirgap {
+				logrus.Debugf("checking airgap bundle matches binary")
+				if err := checkAirgapMatches(airgapBundle); err != nil {
+					return err // we want the user to see the error message without a prefix
+				}
+			}
+
+			if !isAirgap {
```
Comment (sgalsaleh) [inline cmd/installer/cli/install.go:232]: nit:
```suggestion
			else {
```

File: cmd/installer/cli/install.go (hunk 4 above)
Comment (marccampbell) [reply cmd/installer/cli/install.go:221]: why is one better than the other here?

File: cmd/installer/cli/install.go [hunk 6]
```diff
@@ -0,0 +1,1007 @@
...
+
+	return cfg, nil
+}
+
+// DeterminePodAndServiceCIDRS determines, based on the command line flags,
+// what are the pod and service CIDRs to be used for the cluster. If both
+// --pod-cidr and --service-cidr have been set, they are used. Otherwise,
+// the cidr flag is split into pod and service CIDRs.
+func DeterminePodAndServiceCIDRs(cmd *cobra.Command) (string, string, error) {
+	if cmd.Flags().Changed("pod-cidr") && cmd.Flags().Changed("service-cidr") {
```
Comment (sgalsaleh) [inline cmd/installer/cli/install.go:823]: This needs to be `||`
```suggestion
	if cmd.Flags().Changed("pod-cidr") || cmd.Flags().Changed("service-cidr") {
```

File: cmd/installer/cli/install_runpreflights.go [hunk 7]
```diff
@@ -0,0 +1,583 @@
...
+
+	cmd := &cobra.Command{
+		Use:   "run-preflights",
+		Short: "Run install host preflights",
+		PreRunE: func(cmd *cobra.Command, args []string) error {
+			if os.Getuid() != 0 {
+				return fmt.Errorf("run-preflights command must be run as root")
+			}
+
+			// bind proxy flags
```
Comment (sgalsaleh) [inline cmd/installer/cli/install_runpreflights.go:67]: this logic seems duplicated in a few places

File: cmd/installer/cli/install_runpreflights.go [hunk 8]
```diff
@@ -0,0 +1,583 @@
...
+			license, err := getLicenseFromFilepath(licenseFile)
+			if err != nil {
+				return err
+			}
+
+			airgapBundle, err := cmd.Flags().GetString("airgap-bundle")
+			if err != nil {
+				return fmt.Errorf("unable to get airgap-bundle flag: %w", err)
+			}
+			isAirgap := false
```
Comment (sgalsaleh) [inline cmd/installer/cli/install_runpreflights.go:135]: nit: 
```suggestion
			isAirgap := airgapBundle != ""
```

File: cmd/installer/cli/install_runpreflights.go [hunk 9]
```diff
@@ -0,0 +1,583 @@
...
+
+func runHostPreflights(cmd *cobra.Command, provider *defaults.Provider, hpf *v1beta2.HostPreflightSpec, proxy *ecv1beta1.ProxySpec) error {
+	if len(hpf.Collectors) == 0 && len(hpf.Analyzers) == 0 {
+		return nil
+	}
+	pb := spinner.Start()
+
+	skipHostPreflightsFlag, err := cmd.Flags().GetBool("skip-host-preflights")
+	if err != nil {
+		return fmt.Errorf("unable to get skip-host-preflights flag: %w", err)
```
Comment (sgalsaleh) [inline cmd/installer/cli/install_runpreflights.go:445]: needs `pb.CloseWithError()`

File: cmd/installer/cli/install_runpreflights.go [hunk 10]
```diff
@@ -0,0 +1,583 @@
...
+	if output.HasWarn() {
+		s := "preflights"
+		if len(output.Warn) == 1 {
+			s = "preflight"
+		}
+		pb.Warnf("%d host %s warned", len(output.Warn), s)
+
+		noPromptFlag, err := cmd.Flags().GetBool("no-prompt")
+		if err != nil {
+			return fmt.Errorf("unable to get no-prompt flag: %w", err)
```
Comment (sgalsaleh) [inline cmd/installer/cli/install_runpreflights.go:499]: needs `pb.CloseWithError()`

File: cmd/installer/cli/install_runpreflights.go [hunk 11]
```diff
@@ -0,0 +1,583 @@
...
+			// We have warnings but we are not in interactive mode
+			// so we just print the warnings and continue
+			pb.Close()
+			output.PrintTableWithoutInfo()
+			return nil
+		}
+		pb.CloseWithError()
+		output.PrintTableWithoutInfo()
+		if !prompts.New().Confirm("Do you want to continue?", false) {
+			return fmt.Errorf("user aborted")
```
Comment (sgalsaleh) [inline cmd/installer/cli/install_runpreflights.go:511]: needs `pb.CloseWithError()`

File: cmd/installer/cli/materialize.go [hunk 12]
```diff
@@ -0,0 +1,76 @@
...
+
+	cmd := &cobra.Command{
+		Use:   "materialize",
+		Short: "Materialize embedded assets into the data directory",
+		PreRunE: func(cmd *cobra.Command, args []string) error {
+			if os.Getuid() != 0 {
+				return fmt.Errorf("materialize command must be run as root")
+			}
+
+			if cmd.Flags().Changed("data-dir") {
```
Comment (sgalsaleh) [inline cmd/installer/cli/materialize.go:29]: should not check if this changed, just always use it to override runtime config

File: cmd/installer/cli/materialize.go [hunk 13]
```diff
@@ -0,0 +1,76 @@
...
+				dd, err := cmd.Flags().GetString("data-dir")
+				if err != nil {
+					return fmt.Errorf("unable to get data-dir flag: %w", err)
+				}
+
+				runtimeConfig.DataDir = dd
+			}
+
+			if cmd.Flags().Changed("local-artifact-mirror-port") {
```
Comment (sgalsaleh) [inline cmd/installer/cli/materialize.go:38]: should not check if this changed, just always use it to override runtime config

File: cmd/installer/cli/materialize.go [hunk 14]
```diff
@@ -0,0 +1,76 @@
...
+				lap, err := cmd.Flags().GetInt("local-artifact-mirror-port")
+				if err != nil {
+					return fmt.Errorf("unable to get local-artifact-mirror-port flag: %w", err)
+				}
+
+				runtimeConfig.LocalArtifactMirror.Port = lap
+			}
+
+			if cmd.Flags().Changed("admin-console-port") {
```
Comment (sgalsaleh) [inline cmd/installer/cli/materialize.go:47]: should not check if this changed, just always use it to override runtime config

**Review by sgalsaleh — DISMISSED** [review]

File: pkg/k0s/install.go [hunk 15]
```diff
@@ -0,0 +1,32 @@
+package k0s
```
Comment (emosbaugh) [inline pkg/k0s/install.go:1]: this looks unused

Comment (marccampbell) [reply pkg/k0s/install.go:1]: fixed, called it

**Review by sgalsaleh — APPROVED** [review]

**Pull request merged**